src/org/opendatakit/briefcase/ui/reused/AggregateServerConnectionDialogForm.java

```diff
@@ -0,0 +1,275 @@
+    gbc.gridwidth = 3;
+    gbc.weightx = 1.0;
+    gbc.fill = GridBagConstraints.BOTH;
+    dialog.add(panel2, gbc);
+    urlField = new JTextField();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 2;
+    gbc.gridy = 1;
+    gbc.gridwidth = 2;
+    gbc.weightx = 1.0;
+    gbc.anchor = GridBagConstraints.WEST;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(urlField, gbc);
+    final JLabel label1 = new JLabel();
+    label1.setText("URL");
+    gbc = new GridBagConstraints();
+    gbc.gridx = 0;
+    gbc.gridy = 1;
+    gbc.anchor = GridBagConstraints.EAST;
+    panel2.add(label1, gbc);
+    final JPanel spacer5 = new JPanel();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 1;
+    gbc.gridy = 1;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(spacer5, gbc);
+    final JLabel label2 = new JLabel();
+    label2.setText("Username");
+    gbc = new GridBagConstraints();
+    gbc.gridx = 0;
+    gbc.gridy = 2;
+    gbc.anchor = GridBagConstraints.EAST;
+    panel2.add(label2, gbc);
+    final JPanel spacer6 = new JPanel();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 1;
+    gbc.gridy = 2;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(spacer6, gbc);
+    usernameField = new JTextField();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 2;
+    gbc.gridy = 2;
+    gbc.gridwidth = 2;
+    gbc.weightx = 1.0;
+    gbc.anchor = GridBagConstraints.WEST;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(usernameField, gbc);
+    final JLabel label3 = new JLabel();
+    label3.setText("Password");
+    gbc = new GridBagConstraints();
+    gbc.gridx = 0;
+    gbc.gridy = 3;
+    gbc.anchor = GridBagConstraints.EAST;
+    panel2.add(label3, gbc);
+    final JPanel spacer7 = new JPanel();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 1;
+    gbc.gridy = 3;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(spacer7, gbc);
+    final JPanel spacer8 = new JPanel();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 3;
+    gbc.gridy = 0;
+    gbc.fill = GridBagConstraints.VERTICAL;
+    panel2.add(spacer8, gbc);
+    passwordField = new JPasswordField();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 2;
+    gbc.gridy = 3;
+    gbc.gridwidth = 2;
+    gbc.weightx = 1.0;
+    gbc.anchor = GridBagConstraints.WEST;
+    gbc.fill = GridBagConstraints.HORIZONTAL;
+    panel2.add(passwordField, gbc);
+    final JPanel spacer9 = new JPanel();
+    gbc = new GridBagConstraints();
+    gbc.gridx = 3;
+    gbc.gridy = 1;
+    gbc.weighty = 1.0;
+    gbc.fill = GridBagConstraints.VERTICAL;
+    dialog.add(spacer9, gbc);
+  }
+
+  /**
+   * @noinspection ALL
+   */
+  public JComponent $$$getRootComponent$$$() {
+    return dialog;
+  }
+
+  public void onConnect(Runnable callback) {
+    connectButton.addActionListener(__ -> callback.run());
+  }
+
+  public void hideForm() {
+    this.setEnabled(false);
+  }
+
+  public Cursor getSavedCursor() {
```

Comment:
No object outside this class should use thiese methods. These are details that only the view needs to know.

Reply:
Why do we have `onConnect()` and `triggerConnect()` in both the view and the controller?

Reply:
In this case, it feels redundant but it isn't. The view offers an `onConnect()` to let the controller bind to the UI event of clicking the button and the controller offers an `onConnect()` to let third parties bind to the business event of having a new valid Aggregate connection configuration object. The third parties, in this case, are the Pull and Push tabs.